plugins/providers/hyperv/cap/configure_disks.rb

```diff
@@ -0,0 +1,235 @@
+              return false
+            end
+          elsif defined_disk_size < requested_disk_size
+            return true
+          else
+            return false
+          end
+        end
+
+        # Creates and attaches a disk to a machine
+        #
+        # @param [Vagrant::Machine] machine
+        # @param [Kernel_V2::VagrantConfigDisk] disk_config
+        def self.create_disk(machine, disk_config)
+          machine.ui.detail(I18n.t("vagrant.cap.configure_disks.create_disk", name: disk_config.name))
+          disk_provider_config = {}
+          disk_provider_config = disk_config.provider_config[:hyperv] if disk_config.provider_config
+
+          if disk_provider_config
+            disk_provider_config = convert_size_vars!(disk_provider_config)
+          end
+
+          # Get the machines data dir, that will now be the path for the new disk
+          guest_disk_folder = machine.data_dir.join("Virtual Hard Disks")
+
+          if disk_config.file
+            disk_file = disk_config.file
+            LOGGER.info("Disk already defiend by user at '#{disk_file}'. Using this disk instead of creating a new one...")
+          else
+            # Set the extension
+            disk_ext = disk_config.disk_ext
+            disk_file = File.join(guest_disk_folder, disk_config.name) + ".#{disk_ext}"
+
+            LOGGER.info("Attempting to create a new disk file '#{disk_file}' of size '#{disk_config.size}' bytes")
+
+            machine.provider.driver.create_disk(disk_file, disk_config.size, disk_provider_config)
+          end
+
+          disk_info = machine.provider.driver.get_disk(disk_file)
+          disk_metadata = {UUID: disk_info["DiskIdentifier"], Name: disk_config.name, Path: disk_info["Path"]}
+
+          machine.provider.driver.attach_disk(disk_file, disk_provider_config)
+
+          disk_metadata
+        end
+
+        # Converts any "shortcut" options such as "123MB" into its byte form. This
+        # is due to what parameter type is expected when calling the `New-VHD`
+        # powershell command
+        #
+        # @param [Hash] disk_provider_config
+        # @return [Hash] disk_provider_config
+        def self.convert_size_vars!(disk_provider_config)
+          conversion_keys = [:BlockSizeBytes, :LogicalSectorSizeBytes, :PhysicalSectorSizeBytes]
+
+          conversion_keys.each do |k|
+            if disk_provider_config.keys.include?(k)
+              if k.is_a? Integer
+                # Assume it is bytes
+                bytes = disk_provider_config[k]
+              elsif k == :BlockSizeBytes
+                bytes = Vagrant::Util::Numeric.string_to_bytes(disk_provider_config[k])
+              elsif k == :LogicalSectorSizeBytes || k == :PhysicalSectorSizeBytes
+                # Logical and Physical can only be these exact values, so converting it
+                # won't work
+                case disk_provider_config[k]
+                  when "4096MB"
```

Comment:
I'm not super familiar with hyper-v, are these the only valid options? Will more ever be added?

Reply:
It is totally possible that more could be added....but yeah these parameters are literally "only" accepting these two values for `PhysicalSectorSizeBytes` and `LogicalSectorSizeBytes` right now, and any other value ends up being an exception: 

> Accepted values: | 512, 4096

https://docs.microsoft.com/en-us/powershell/module/hyper-v/new-vhd?view=win10-ps

Which is why I didn't use the Numeric class here to convert them...they won't end up "exactly" those values so it ends up with a powershell exception.

This entire method is unfortunately a hack to get Vagrant to properly pass along these parameters into the New-VHD powershell command :cry: They're technically expecting a `UInt32`, but they expect units to go along with it, but it also can't be a string like `"129MB"`..it has to be a literal like `-PhysicalSectorSizeBytes 512mb`.

Reply:
Now that I re-read the docs, it looks like maybe we could just delete these switch statements, and expect people to just set it to 512 or 4096. If they don't set it properly, they will see the error from Hyper-V :shrug: Then the only thing we should be converting is BlockSizeBytes.